### PR TITLE
net: wifi: Introduce Wi-Fi network manager's support

### PR DESCRIPTION
**Pull request by krish2718**
This introduces support for Wi-Fi network managers in Zephyr. The
    motivation is for the Wi-Fi management layer to work with both
    Network managers and offloaded Wi-Fi drivers. The device driver
    decides which one to use.
    
    network manager : Apps -> Wi-Fi Mgmt -> Network Manager -> Wi-Fi
    interface
    
    offloaded       : Apps -> Wi-Fi Mgmt -> Wi-Fi offloaded interface
    
    Support for multiple network managers has been added, each device can
    choose its own network manager and there can be mix and match:
    
      wlan0 - Offloaded
      wlan1 - Network manager 1
      wlan2 - Network manager 2

### CODE REVIEW

File: include/zephyr/net/net_if.h [hunk 1]
```diff
@@ -2535,6 +2535,13 @@ bool net_if_is_suspended(struct net_if *iface);
  */
 bool net_if_is_wifi(struct net_if *iface);
 
+/**
+ * @brief Get first Wi-Fi network interface.
+ *
+ * @return Pointer to network interface, NULL if not found.
+ */
+struct net_if *net_if_get_first_wifi(void);
```
Comment (rlubos) [inline include/zephyr/net/net_if.h:2543]: I think commits:
* `net: wifi: Add support to get and set Wi-Fi as default interface`
*  `net: wifi: Use only Wi-Fi interfaces`

could be a part of #58869 - they don't seem related to the rest of `wifi_mgmt` changes in here, and are a nice supplement to changes in #58869, as they actually show the reason for those changes.

File: include/zephyr/net/ethernet.h [hunk 2]
```diff
@@ -352,8 +352,11 @@ enum ethernet_if_types {
 	/** IEEE 802.3 Ethernet (default) */
 	L2_ETH_IF_TYPE_ETHERNET,
 
-	/** IEEE 802.11 Wi-Fi*/
+	/** Offloaded IEEE 802.11 Wi-Fi (default Wi-Fi) */
 	L2_ETH_IF_TYPE_WIFI,
+
+	/** Native IEEE 802.11 Wi-Fi */
+	L2_ETH_IF_TYPE_WIFI_NATIVE,
```
Comment (rlubos) [inline include/zephyr/net/ethernet.h:359]: After looking at the logic in commit `net: wifi: Add support to handle Wi-Fi native to Wi-Fi management` - is this separate `WIFI_NATIVE` type even needed? It doesn't seem to rely on the interface type, but rather have a nice fallback mechanism - if there's no "shared" implementation found, it falls back to offloaded one.

File: subsys/net/l2/wifi/Kconfig [hunk 3]
```diff
@@ -27,3 +27,20 @@ config WIFI_MGMT_RAW_SCAN_RESULTS_ONLY
 	  to the application.
 
 endif # WIFI_MGMT_RAW_SCAN_RESULTS
+
+config WIFI_NATIVE
```
Comment (rlubos) [inline subsys/net/l2/wifi/Kconfig:31]: IMO (but it's not only my opinion) the "wifi_native" nomenclature is a bit confusing. Especially "native Wi-Fi stack"- it suggests that we have a complete Wi-FI stack (including MAC) for Zephyr. In reality though, if I understood the intent of this PR correctly, it's more about providing a framework to register a Wi-Fi supplicant instance, that can be shared among multiple interfaces.

File: include/zephyr/net/wifi_native.h [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+#endif
+
+/**
+ * @brief Wi-Fi Native Managed Interface
+ */
+struct wifi_native_mgd_iface {
+	/** Interface */
+	struct net_if *iface;
+	/** Wi-Fi Management operations */
+	const struct wifi_mgmt_ops *ops;
```
Comment (rlubos) [inline include/zephyr/net/wifi_native.h:40]: Shouldn't `struct wifi_mgmt_ops` be rather part of `struct wifi_native_stack`? My understanding is that we want to share a common implementation across all managed interfaces?

File: subsys/net/l2/wifi/wifi_native.c [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	SYS_SLIST_FOR_EACH_CONTAINER(&stack->mgd_ifaces, mgd_iface, node) {
+		if (mgd_iface->iface == iface) {
+			return -EEXIST;
+		}
+	}
+
+	mgd_iface = k_malloc(sizeof(*mgd_iface));
+	if (!mgd_iface) {
+		return -ENOMEM;
+	}
```
Comment (rlubos) [inline subsys/net/l2/wifi/wifi_native.c:54]: Do we really need to involve heap here, we usually aim to avoid that in the net stack. I'd prefer to see a preallocated array in `struct wifi_native_stack` instead, especially that such information on what interface will be managed by which supplicant could be easily determined build-time, we can't create network interfaces at runtime in Zephyr.

**Review by rlubos — COMMENTED** [review]
I think we need to rethink `wifi_native` nomenclature here. Plus I'd really like to see some documentation (and maybe tests for the new framework), as it bring extra complexity to the Wi-Fi subsystem that may not be very intuitive for newcomers.

File: include/zephyr/net/ethernet.h (hunk 2 above)
Comment (krish2718) [reply include/zephyr/net/ethernet.h:359]: Well, if we have a `type` to convey, then this `enum` is needed to convey. It can also be used as a check for Native stack to ensure given interface is `native`.

File: subsys/net/l2/wifi/Kconfig (hunk 3 above)
Comment (krish2718) [reply subsys/net/l2/wifi/Kconfig:31]: Agreed, even I was thinking the same after publishing this PR, but thought will wait for others to chime in. Should we just call this `supplicant` support? but it might be mistaken with `wpa_supplicant`.

File: subsys/net/l2/wifi/wifi_native.c [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wifi_native, CONFIG_WIFI_NATIVE_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/wifi_native.h>
+
+sys_slist_t wifi_native_stacks = SYS_SLIST_STATIC_INIT(&wifi_native_stacks);
+
+void wifi_native_register_stack(struct wifi_native_stack *stack)
```
[the remark this answers is not in the document]
Comment (krish2718) [reply subsys/net/l2/wifi/wifi_native.c:15]: Had the same concern here https://github.com/zephyrproject-rtos/zephyr/pull/59249#discussion_r1230814631

File: include/zephyr/net/wifi_native.h (hunk 4 above)
Comment (krish2718) [reply include/zephyr/net/wifi_native.h:40]: Yes, I can move it there.

**Comment by krish2718** [comment]
> I think we need to rethink `wifi_native` nomenclature here. Plus I'd really like to see some documentation (and maybe tests for the new framework), as it bring extra complexity to the Wi-Fi subsystem that may not be very intuitive for newcomers.

I am writing docs and unit tests already, but raised this as draft to get early feeedback.

File: subsys/net/l2/wifi/wifi_native.c (hunk 5 above)
Comment (krish2718) [reply subsys/net/l2/wifi/wifi_native.c:54]: I will convert linked lists for `stacks` and `mgd_ifaces` to arrays with configuration options.

File: include/zephyr/net/ethernet.h (hunk 2 above)
Comment (krish2718) [reply include/zephyr/net/ethernet.h:359]: I have moved the two commits over to the other PR, and I will remove the WIFI_NATIVE altogether.

File: include/zephyr/net/wifi_mgmt.h [hunk 7]
```diff
@@ -332,6 +320,21 @@ struct net_wifi_mgmt_offload {
...
+	 * A network device should indeed provide a pointer on such
+	 * net_if_api structure. So we make current structure pointer
+	 * that can be casted to a net_if_api structure pointer.
+	 */
+#ifdef CONFIG_WIFI_USE_NATIVE_NETWORKING
+	struct ethernet_api wifi_iface;
+#else
+	struct offloaded_if_api wifi_iface;
+#endif
+	const struct wifi_mgmt_ops *const wifi_api;
```
[the remark this answers is not in the document]
Comment (krish2718) [reply include/zephyr/net/wifi_mgmt.h:335]: `wifi_mgmt_api`?

[the remark this answers is not in the document]
Comment (rlubos) [reply include/zephyr/net/wifi_mgmt.h:335]: Sounds good

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 8]
```diff
@@ -12,6 +12,15 @@ LOG_MODULE_REGISTER(net_wifi_mgmt, CONFIG_NET_L2_WIFI_MGMT_LOG_LEVEL);
 #include <zephyr/net/net_core.h>
 #include <zephyr/net/net_if.h>
 #include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/wifi_native.h>
+static const struct wifi_mgmt_ops *const get_wifi_api(struct net_if *iface)
```
Comment (rlubos) [inline subsys/net/l2/wifi/wifi_mgmt.c]: Please add an empty line here.

File: include/zephyr/net/wifi_nm.h [hunk 9]
```diff
@@ -0,0 +1,100 @@
+/** @file
+ * @brief Wi-Fi Network manager API
+ *
+ * This file contains the Wi-Fi network manager API. These APIs are used by the
+ * any network management application to register as a Wi-Fi network manager.
+ */
+
+/*
+ * Copyright (c) Nordic Semiconductor ASA.
```
Comment (rlubos) [inline include/zephyr/net/wifi_nm.h:9]: Year missing.

File: subsys/net/l2/wifi/wifi_nm.c [hunk 10]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Nordic Semiconductor ASA.
```
Comment (rlubos) [inline subsys/net/l2/wifi/wifi_nm.c:2]: Year missing.

File: include/zephyr/net/wifi_nm.h [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+		.ops = _ops,				\
+		.mgd_ifaces = { NULL },		\
+	}
+
+/**
+ * @brief Get a Network manager instance for a given name
+ *
+ * @param name Name of the Network manager instance
+ *
+ * @return struct wifi_nm_instance*  Pointer to Network manager instance
```
Comment (rlubos) [inline include/zephyr/net/wifi_nm.h:59]: There's no need to document returned type, it won't render nicely. Here and below.

File: include/zephyr/net/wifi_nm.h [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+ */
+struct wifi_nm_instance *wifi_nm_get_instance_iface(struct net_if *iface);
+
+/**
+ * @brief Register a managed interface
+ *
+ * @param nm Pointer to Network manager instance
+ * @param iface Managed interface
+ *
+ * @return int 0 for OK; -EINVAL for invalid parameters; -EEXIST if interface is already registered
```
Comment (rlubos) [inline include/zephyr/net/wifi_nm.h:78]: If you enumerate possible return values, it'd be better to use multiple `@retval` entries.

File: subsys/net/l2/wifi/wifi_mgmt.c [hunk 13]
```diff
@@ -28,9 +38,8 @@ static int wifi_connect(uint32_t mgmt_request, struct net_if *iface,
 	struct wifi_connect_req_params *params =
 		(struct wifi_connect_req_params *)data;
 	const struct device *dev = net_if_get_device(iface);
-	struct net_wifi_mgmt_offload *off_api =
-		(struct net_wifi_mgmt_offload *) dev->api;
-	const struct wifi_mgmt_ops *const wifi_mgmt_api = off_api ? off_api->wifi_mgmt_api : NULL;
+
```
Comment (rlubos) [inline subsys/net/l2/wifi/wifi_mgmt.c:41]: No need for this empty line.

File: tests/net/wifi/wifi_nm/CMakeLists.txt [hunk 14]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(wifi)
```
Comment (rlubos) [inline tests/net/wifi/wifi_nm/CMakeLists.txt:5]: `wifi_nm`

File: tests/net/wifi/wifi_nm/src/main.c [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+	int ret;
+	struct wifi_nm_instance *nm = wifi_nm_get_instance("test");
+
+	zassert_equal(nm->ops, &wifi_nm_test_ops,
+		      "Invalid wifi nm ops");
+
+	ret = request_scan();
+	zassert_equal(ret, 0, "Scan request failed");
+	zassert_true(wifi_offload_op_called, "Scan callback not called");
```
Comment (rlubos) [inline tests/net/wifi/wifi_nm/src/main.c:168]: Wasn't offloaded case tested in the case above?

**Review by rlubos — COMMENTED** [review]
I've left a few more nits, but overall IMO it looks pretty good, the intent of this feature is much clearer now.

File: tests/net/wifi/wifi_nm/src/main.c [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	return 0;
+}
+
+ZTEST(net_wifi, test_wifi_offload)
+{
+
+	int ret;
+#ifdef CONFIG_WIFI_NM
```
[the remark this answers is not in the document]
Comment (krish2718) [reply tests/net/wifi/wifi_nm/src/main.c:142]: Well, just in case someone wants to disable and check, but I can remove.

File: tests/net/wifi/wifi_nm/src/main.c (hunk 15 above)
Comment (krish2718) [reply tests/net/wifi/wifi_nm/src/main.c:168]: Yes, this is offloaded when `nm` is registered but there are no managed interfaces.

**Review by rlubos — DISMISSED** [review]
LGTM

**Review by rlubos — APPROVED** [review]

**Comment by ssharks** [comment]
LGTM, I'm not going to give an official approval at this moment as I think it is too important to merge just with 2 green.

**Review by carlescufi — APPROVED** [review]
